Support for non-clients, NVidia shield, dynamic grouping, extra metad…

JesseWebDotCom opened the pull request:
This is my first contribution so forgive me if I misstep.

A full breakdown of the updates are here:
[](https://github.com/JesseWebDotCom/home-assistant-configuration/blob/master/docs/media_player_plexdevices.md)

If you agree with the direction and or changes, I can start drafting the updated version of:
https://home-assistant.io/components/media_player.plex/

Let me know if there is an official process for updating documentation
Thanks

JesseWebDotCom committed:
Support for non-clients, NVidia shield, dynamic grouping, extra metadata, and more

homeassistant commented:
Hi @JesseWebDotCom, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#6054).

Once you do that we will be able to review and accept this pull request.

Thanks!

JesseWebDotCom commented:
I can't sign the CLA - the web page spinner never stops (tried from multiple browsers, clients, and networks).

JesseWebDotCom committed:
Fixed import orderdering, line lengths, and comment violations

homeassistant commented:
Hi @JesseWebDotCom, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#6054).

Once you do that we will be able to review and accept this pull request.

Thanks!

JesseWebDotCom committed:
Local player check and season fixes

homeassistant commented:
Hi @JesseWebDotCom, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#6054).

Once you do that we will be able to review and accept this pull request.

Thanks!

JesseWebDotCom committed:
Honor entity_namespace when using custom entity ids

homeassistant commented:
Hi @JesseWebDotCom, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#6054).

Once you do that we will be able to review and accept this pull request.

Thanks!

JesseWebDotCom committed:
Semi compatibility with channels, force controls option added

homeassistant commented:
Hi @JesseWebDotCom, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#6054).

Once you do that we will be able to review and accept this pull request.

Thanks!

robbiet480 commented:
@JesseWebDotCom Can you provide a full list of changes?

JesseWebDotCom commented:
My original documentation is here:
https://github.com/JesseWebDotCom/home-assistant-configuration/blob/master/docs/media_player_plexdevices.md

change log (hopefully that’s everything):
* Improved compatibility
	* Added support for non-clients
		* remote users
		* PlexConnect Apple TV’s
		* Other non-controllable clients
	* Improved support for channels (displays and is controllable providing the channel session closely resembles a movie or TV session)
* Improved controls
	* Added volume controls (volume up/down, mute)
		* Faking starting volume (as 100) as plexapi does not expose current volume
		* Faking muting as plexapi does not expose mute state
	* Added Off
		* Faking by stopping media as plexapi does not provide an off method
	* Added progress bar
	* Restrict controls based on identified compatibility issues:
		* Remove controls if client is local (i.e. loopback with no controllable IP)
		* Remove mute for Nvidia shields (they do not support volume lower than 2)
	* Added ability (enabled through optional config) to avoid control restriction and always show all controls
* Improved metadata
	* Display library name in “app_name” attribute
	* Display movie name as “name (year)”, ex: “Blair Witch (2016)”
	* Display TV show and episode number with leading 0 (ex. S02 instead of S2)
	* Display music artist
	* All music attributes (media artist, album name, album artist, track number)
		* Fallback to Album artist is media artist is blank
	* Media position and duration attributes
	* Added fallback to display TV season number when API seasons() call fails
	* Added ability (enabled through optional config) to display the thumbnail for the playing TV episode as opposed to the show
	* Display art if thumbnail is non-existent (useful for tv/movies without posters)
* Added ability (enabled through optional config) to have entity ID’s named media_player.{plex player id} instead of media_player.{plex player name}.
* Added PlayMedia service abilities to make a plex client play a movie, tv show, or music playlist
* (enabled through optional config) Added dynamic “active” and “inactive” groups
	* I’d like to establish a pattern here for automated grouping and to avoid conflicts with manual grouping.  I chose to prefix my automated group names with underscores.
	* There has to be some problem with how I created the groups as when I run this “{{ states.group._plex_devices_disconnected.attributes.entity_id | length }}” I get this “Error rendering template: SecurityError: access to attribute '_plex_devices_disconnected' of 'DomainStates' object is unsafe.”
* Added optional config parameters (not sure I did this right but it works) to do the following:
	* 'include_non_clients’: includes non-controllable clients (ex. PlexConnect Apple TV’s), defaults to false (for backwards compatibility)
	* 'use_episode_art’: Use episode art instead of show art, defaults to false (for backwards compatibility)
	* 'use_dynamic_groups’: Automatically group devices into active and inactive groups, defaults to false (for backwards compatibility)
	* 'use_custom_entity_ids’: Name entities by device id (less ambiguous, more predictable names), defaults to false (for backwards compatibility)
	* 'show_all_controls’: Always show all controls instead of only displaying ones within client capabilities, defaults to false (for backwards compatibility)
* Misc improvements
	* Improved refresh speed by querying sessions first, devices second
	* Improved state accuracy by forcing idle status for old/invalid sessions
	* Improved error avoidance by checking if object’s are not none and if they are callable
	* Fixed bug where TV season would display even though device has gone idle
	* Removed unnecessary code:
		* Removed “return STATE_UNKNON” code as it was unreachable
		* Removed “import Show” references as we already know the media type

JesseWebDotCom committed:
media_position_updated_at fix - only update when media is playing

JesseWebDotCom committed:
Fix: controls now work as expected on 1) casted sessions and 2) client sessions when client and PMS reside on the same sever

JesseWebDotCom commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +428,302 @@ def _active_media_plexapi_type(self):
+            # with a loopback address, the device must be local or casting
+            for client in self.device.server.clients():
+                if "127.0.0.1" in client.baseurl:
+                    if client.machineIdentifier == self.device.machineIdentifier:
```
This is more readable on one line but I'll change it if you need me to.

robbiet480 replied:
Please do, as we have strict requirements on PEP8 compliance.

JesseWebDotCom replied:
ok, so I'm not sure how to fix this as I got an auto message each way I went

JesseWebDotCom committed:
Made PEP8 compliant

JesseWebDotCom commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +428,303 @@ def _active_media_plexapi_type(self):
+            for client in self.device.server.clients():
+                if "127.0.0.1" in client.baseurl:
+                    if client.machineIdentifier == \
+                        self.device.machineIdentifier:
```
Boo, will fix (hopefully)

JesseWebDotCom committed:
Made PEP8 compliant

JesseWebDotCom committed:
Made PEP8 compliant, hopefully

robbiet480 commented:
@JesseWebDotCom You can use `pylint` and `flake8` locally before pushing changes to ensure you are PEP8 compliant.

JesseWebDotCom commented:
Thanks.  Already use pylint in Atom.io combined with Atom beautify using yapf.  It usually fixes everything automatically.  Just added flake8 for good measure.

JesseWebDotCom commented:
Is there any review process for documentation as I will need to update this page once you approve my code: https://home-assistant.io/components/media_player.plex/

In the interim, I'll start drafting it offline

fabaff commented:
Docs: https://github.com/home-assistant/home-assistant.github.io/pull/2094

JesseWebDotCom committed:
Better Tivo compatibility

JesseWebDotCom committed:
Added frozen detection and remediation

JesseWebDotCom committed:
PlayMedia EPISODE now supports season number and episode number (instead of episode index)

JesseWebDotCom committed:
Fix: Dynamic groups now exclude hidden devices

JesseWebDotCom committed:
Fix: Dynamic groups now exclude hidden devices

JesseWebDotCom commented on homeassistant/components/media_player/plex.py:
```diff
@@ -136,12 +209,61 @@ def update_devices():
+>>>>>>> 4201005ca8ff3f7b584e5cbb9f3464f65c8ce2a9
+
+            # set groups with updated memberships
+            set_group_members(hass, GROUP_ACTIVE_DEVICES,
```
Ignore, I pushed a new update (fix3) that works and doesn't include the junk text above

JesseWebDotCom commented:
Any word on when this and the documentation (home-assistant/home-assistant.github.io#2094) will get officially incorporated?

arsaboo commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +271,225 @@ def update_sessions():
+        if self.session and self.session.player:
+            if ((frozen_seconds >
+                 int(self.optional_config[CONF_MAX_FROZEN_PAUSED])) and
+                (self.session.player.state == 'paused' or
```
Need one space here.

arsaboo commented:
@JesseWebDotCom Looks like the build is failing.....there are some simple errors that can be fixed.

arsaboo commented on homeassistant/components/media_player/plex.py:
```diff
@@ -79,22 +145,28 @@ def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+
 
+def set_group_members(hass, group_entity_id, member_entity_id_list):
+    """Creates group if doesn't exist and sets memberships"""
```
Needs to end in a period
`"""Creates group if doesn't exist and sets memberships."""`

arsaboo commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +271,225 @@ def update_sessions():
+                        self.name.lower().replace('-', '_'))
+
+    def refresh(self, device, session):
+        """Refresh key device data"""
```
`"""Refresh key device data."""`

arsaboo commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +271,225 @@ def update_sessions():
+                self._make = make
+
+    def force_idle(self):
+        """Force client to idle"""
```
`"""Force client to idle."""`

arsaboo commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +271,225 @@ def update_sessions():
+
+    @property
+    def app_name(self):
+        """Library name of playing media"""
```
"""Return the library name of playing media."""

arsaboo commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +512,381 @@ def _active_media_plexapi_type(self):
+                    return None
+
+    def local_client_control_fix(self):
+        """Detects if local client and adjusts url to allow control"""
```
`"""Detect if local client and adjust url to allow control."""`

arsaboo commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +512,381 @@ def _active_media_plexapi_type(self):
+
+    def _get_episode(self, library_name, show_name, season_number,
+                     episode_number):
+        """Finds TV episode and returns a Plex media object"""
```
`"""Finds TV episode and return a Plex media object."""`

arsaboo commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +512,381 @@ def _active_media_plexapi_type(self):
+        return target_episode
+
+    def _client_play_media(self, media, **params):
+        """Instructs Plex client to play a piece of media."""
```
`"""Instruct Plex client to play media."""`

JesseWebDotCom committed:
Implemented arsaboo suggested formatting

JesseWebDotCom commented:
I implemented the changes but the build is still failing.  How can I see for myself exactly where its failing and what to do to fix it?

JesseWebDotCom committed:
Implemented pylint command line suggested fixes

arsaboo commented:
You can see in the Travis build report...it indicates the errors and how to correct them.

JesseWebDotCom committed:
Implemented Travis CI build suggested fixes

arsaboo commented on homeassistant/components/media_player/plex.py:
```diff
@@ -4,35 +4,68 @@
+import requests
+from requests.packages.urllib3.exceptions import InsecureRequestWarning
+
+import asyncio
```
Should come before `import requests`

JesseWebDotCom replied:
Can you include a screenshot of how you are seeing the errors?  When I click on the build jobs, I see a bombardment of lines, what look like sql queries, errors that don't pertain to the code, etc?  Thanks.

JesseWebDotCom committed:
Sorted Imports using Importanize

JesseWebDotCom committed:
Grouped request imports

JesseWebDotCom commented:
Is this coveralls thing something I need to deal with?

arsaboo commented:
Start by adding `plexapi==2.0.2` to the file requirements_all.txt. If that does not work, you may have to write tests.

JesseWebDotCom commented:
That is currently in the requirements file and I how no idea how I would write tests.  How do you interpret these coverall results anyways?  Is there someone else who can help me get this over the line?

pvizeli commented:
For that should be a part of official things, you need change some hacks:
- Do not setup a plex Server / is not a part of hass.
- Remove your group stuff
- Do not fake stuff they is not supported by API
- Make no blocking stuff inside asyncio loop

pvizeli commented:
For that should be a part of official things, you need change some hacks:
- Do not setup a plex Server / is not a part of hass.
- Remove your group stuff
- Do not fake stuff they is not supported by API
- Make no blocking stuff inside asyncio loop

pvizeli reviewed (dismissed):
For that should be a part of official things, you need change some hacks:
- Do not setup a plex Server / is not a part of hass.
- Remove your group stuff
- Do not fake stuff they is not supported by API
- Make no blocking stuff inside asyncio Loop
- Do not change file permissions

JesseWebDotCom commented:
@pvizeli - Thanks, I'll review your changes.  I'm not yet in agreement with 1) removing the group stuff and 2) faking stuff.

For #1, I want to establish a pattern where components can dynamically group entities.  In this case I find it extremely helpful to have a group is active and inactive devices.  I don't need to create or manage any groups for plex devices as they are already done.  So I'll stop here and ask, are you in agreement with the need for dynamic groups (I believe HA does this in other cases), if not why, but if so then how are you recommending me achieve them in this case?

For #2, this "faking" brings a decent set of missing functionality.  I'd rather have it off by default and end user configurable to enable then to completely remove.

I didn't understand your other notes, ex. don't setup a plex server, don't block inside async loo, and con't change file permissions.  If you already changed the code to fix these things, I'll review and try to spot/understand what you mean.

Thanks for taking the time to do this.  I've put a lot of time and effort to bring something I find useful to the community.  This last bit of help from others is needed to make that happen and appreciated.

Hellowlol commented:
@JesseWebDotCom if you feel anything is missing from plexapi send a pr ;)

JesseWebDotCom commented:
Guys - I feel we have gotten to the point of diminishing returns and the code is ready for release.  I ask that you approve the code as is.  Otherwise I will keep this as a custom component for my own personal use.  That would be a shame as this will bring great value for HA Plex users - I just can't spend any more time on this.

robbiet480 commented:
Taking a look at this now.

As for the permissions problem @JesseWebDotCom, that's an issue that happened as you were working on this locally. The file permission you assign transfers into Git. I can get it fixed for you on my computer so we can consider that issue fixed.

Will comment on other issues shortly.

robbiet480 commented:
@pvizeli Seems when you saw `setup_plexserver` you thought that this meant an entire Plex server was being started. It's actually just a poorly named function. A long time ago this platform was called `plexserver` and it's a leftover from that.

robbiet480 commented:
Okay so last issue I believe is the blocking inside async loops. @pvizeli can you point out where you are talking about, I can't find it with a quick look.

Hoping we can get this into 0.40 but can't make promises at this point.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +286,225 @@ def update_sessions():
+
+        if self.session and self.session.player:
+            player_id = self._convert_na_to_none(
+                self.session.player.machineIdentifier)
```
Is this doing a network request? We should not do any I/O when fetching properties. Instead, all data should be cached inside the `update` method.

JesseWebDotCom replied:
Roger that - I wasn't aware that wasn't allowed (makes sense to grab everything up front).  I will update (this and the others you mentioned).

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +286,225 @@ def update_sessions():
+            if self.session:
+                if self.session.librarySectionID:
+                    return self._convert_na_to_none(
+                        self.session.server.library.sectionByID(
```
Same

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +286,225 @@ def update_sessions():
         """Return the session, if any."""
-        return self.plex_sessions.get(self.device.machineIdentifier, None)
+        if self.device is not None:
+            session = self.plex_sessions.get(self.device.machineIdentifier,
```
Same

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +527,383 @@ def _active_media_plexapi_type(self):
     def media_content_type(self):
         """Content type of current playing media."""
-        if self.session is None:
+        if self.state not in (STATE_IDLE, STATE_OFF):  # true or frozen state
```
Consider using a guard clause:

```python
if self.state in (STATE_IDLE, STATE_OFF):
    return None

if not self.session:
    return None

media_type = …
```

balloob replied:
We also seem to check these two statements regularly. Would it make sense to put it in a property  ? Or, just resolve all these things inside the `update` method.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +527,383 @@ def _active_media_plexapi_type(self):
+    def media_artist(self):
+        """Artist of current playing media, music track only."""
+        if self.state not in (STATE_IDLE, STATE_OFF):  # true or frozen state
+            if self.media_content_type is MEDIA_TYPE_MUSIC:
```
Use `==` for equal comparison.

JesseWebDotCom replied:
Are you referring to the "is MEDIA_TYPE MUSIC"?  If so, can you explain when I should use is vs equals?

arsaboo replied:
You can use something like `self.media_content_type == MEDIA_TYPE_MUSIC`. See [here](https://github.com/home-assistant/home-assistant/blob/5f0b2a7d15df474c47c73c2ddac3fa9e8bd4f332/homeassistant/components/media_player/gstreamer.py#L91) for example.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +527,383 @@ def _active_media_plexapi_type(self):
+
+                if thumb_url:
+                    thumb_url = self.session.server.url(thumb_url)
+                    thumb_response = requests.get(thumb_url, verify=False)
```
If we need to do web requests to validate a property, this should happen inside the `update` method.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +527,383 @@ def _active_media_plexapi_type(self):
+        if self.state not in (STATE_IDLE, STATE_OFF):  # true or frozen state
+            if self.media_content_type is MEDIA_TYPE_TVSHOW:
+                if self.session is not None:
+                    if callable(self.session):
```
Session is not used as a callable anywhere else, are you sure this should not be `seasons`?

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +527,383 @@ def _active_media_plexapi_type(self):
     def supported_features(self):
         """Flag media player features that are supported."""
-        return SUPPORT_PLEX
+        if self.state not in (STATE_IDLE, STATE_OFF):  # true or frozen state
```
Every if check with no else clause or a limited one (like return None / 0) should become a guard clause.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +527,383 @@ def _active_media_plexapi_type(self):
+    @property
+    def volume_level(self):
+        """Return the volume level of the client (0..1)."""
+        if self.state not in (STATE_IDLE, STATE_OFF):  # true or frozen state
```
We should never have two embedded `if` statements. Instead they should be merged into 1 if statement.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +527,383 @@ def _active_media_plexapi_type(self):
+            self.device.skipPrevious(self._active_media_plexapi_type)
+
+    @asyncio.coroutine
+    def async_play_media(self, media_type, media_id, **kwargs):
```
The things you call are not async safe. Please overwrite `play_media` instead.

balloob reviewed (changes requested):
This platform will still need some work.

It is not allowed to do any network requests or other I/O inside async methods or inside properties. All values should be fetched and cached inside the `update` method.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -79,22 +161,28 @@ def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+    setup_plexserver(host, token, hass, optional_config, add_devices_callback)
+
 
+def set_group_members(hass, group_entity_id, member_entity_id_list):
```
This is not allowed. This is not part of how groups work in Home Assistant. Please remove.

JesseWebDotCom replied:
See comment below

JesseWebDotCom commented:
@ballob - can you provide me a way to do what I'm trying to do with the groups (be able to dynamically see which devices are active and which are not without having to manually manage group memberships or have a few automations)?  What are your thoughts on dynamic grouping by components (ex the chromecast component could have dynamic groups for playing vs not)?

JesseWebDotCom commented:
Thanks all - as a newbie, this was  all a little overwhelming.  Some of your latest comments helped make things real for me - so I'll update some things tonight.  I'll clearly note which comments I addressed and which I didn't (probably because I didn't understand what was intended).

balloob commented:
You have to work directly with the group component but I would suggest that we leave that for another PR. This PR is already getting very big and there will always be time to do more PRs after this one 👍

JesseWebDotCom commented:
@balloob - I'll remove the group features but please send me 1) an example of how to do it correctly and 2) the pattern you'd like to use for dynamic groups (I chose prefixing group names with underscores).

JesseWebDotCom committed:
Removed dynamic groups, network calls from properties, and other cleanup

JesseWebDotCom commented:
I believe I fixed all issues except for the permissions issue Robbie said he would address.  Please review and let me know what's left.  BTW - that was a lot of work (at least for me)

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -136,12 +214,30 @@ def update_devices():
+        # add devices with a session and no client (ex. PlexConnect Apple TV's)
+        if optional_config[CONF_INCLUDE_NON_CLIENTS]:
+            for machine_identifier, session in plex_sessions.items():
+                if machine_identifier not in plex_clients:
```
You need to put `and machine_identifier is not None` - this is causing duplicate entities for same players.

Teagan42 commented:
There is one problem I am seeing - I am now getting duplicate "clients". This is something I had to address in my PR (didn't realize you had this open when I made mine last night).

My PR https://github.com/home-assistant/home-assistant/pull/6530

What happens is, there are sessions, with no machineIdentifier which causes it to create a non-client entity.

This right here fixes the duplicate entities.
https://github.com/home-assistant/home-assistant/pull/6530/files#diff-c09045bdd393c9276405475cce2ec8eaR162

Added a comment to where you should insert similar logic to fix it.

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -7,32 +7,77 @@
 
 PLEX_CONFIG_FILE = 'plex.conf'
 
+GROUP_ACTIVE_DEVICES = 'group._plex_devices_active'
```
This is no longer used after your last commit

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -7,32 +7,77 @@
 PLEX_CONFIG_FILE = 'plex.conf'
 
+GROUP_ACTIVE_DEVICES = 'group._plex_devices_active'
+GROUP_INACTIVE_DEVICES = 'group._plex_devices_inactive'
```
This is no longer used after your last commit

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
+            features = (SUPPORT_PAUSE | SUPPORT_PLAY | SUPPORT_STOP |
+                        SUPPORT_TURN_OFF)
+        # Default to show all controls
+        elif self.device:
```
This case is wrong:

should be:

```
        if self.device and 'playback' in self.device.protocolCapabilities:
            return (SUPPORT_PAUSE | SUPPORT_PREVIOUS_TRACK |
                        SUPPORT_NEXT_TRACK | SUPPORT_STOP |
                        SUPPORT_VOLUME_SET | SUPPORT_PLAY | SUPPORT_SEEK |
                        SUPPORT_TURN_OFF | SUPPORT_VOLUME_MUTE)
        else:
          features = None
```

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
         """Set volume level, range 0..1."""
-        self.device.setVolume(int(volume * 100),
-                              self._active_media_plexapi_type)
+        if self.device:
```
`self.device` is not enough to know if it supports volume controls - needs to include `playback` in the `device.protocolCapabilities` (which still doesn't tell you if it supports volume controls.

JesseWebDotCom replied:
Then there is nothing we can do?

JesseWebDotCom replied:
Help me understand how you think this should be implemented.  As is, I am able to control volume on my Nvidia Shield and my iPhone.

JesseWebDotCom replied:
I added the following to all control commands:
`if self.device and 'playback' in self._device_protocol_capabilities:`

I'll commit later after I test

Teagan42 replied:
These are the capabilities the device property may have:
`timeline,playback,navigation,mirror,playqueues`

`playback` is used for: playMedia, stop, skip-forward (fast forward), skip-backward (rewind), skipNext (next track), skipPrevious (previous track), seekTo (go to time X in current playing media), refresh play queue, play, pause, stop, setStream, repeat, shuffle

You should look at the library we're using https://github.com/pkkid/python-plexapi

So basically, almost ALL of our support flags have to do with `playback`. So, `if 'playback' in self.device.protocolCapabilities` is the check you need to determine if it supports ANYTHING or NOTHING.

JesseWebDotCom replied:
I also cached protocol capabilities in my refresh method to avoid network calls in properties

Teagan42 replied:
`protocolCapabilities` is not a network call inside plexapi as far as I can tell.

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
     def media_play(self):
         """Send play command."""
-        self.device.play(self._active_media_plexapi_type)
+        if self.device:
```
`self.device` is not enough to know if it supports remote commands - needs to include `playback` in the `device.protocolCapabilities` (which still doesn't tell you if it supports volume controls.

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
     def media_pause(self):
         """Send pause command."""
-        self.device.pause(self._active_media_plexapi_type)
+        if self.device:
```
`self.device` is not enough to know if it supports remote commands - needs to include `playback` in the `device.protocolCapabilities` (which still doesn't tell you if it supports volume controls.

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
     def media_stop(self):
         """Send stop command."""
-        self.device.stop(self._active_media_plexapi_type)
+        if self.device:
```
`self.device` is not enough to know if it supports remote commands - needs to include `playback` in the `device.protocolCapabilities` (which still doesn't tell you if it supports volume controls.

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
     def media_next_track(self):
         """Send next track command."""
-        self.device.skipNext(self._active_media_plexapi_type)
+        if self.device:
```
`self.device` is not enough to know if it supports remote commands - needs to include `playback` in the `device.protocolCapabilities` (which still doesn't tell you if it supports volume controls.

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
     def media_previous_track(self):
         """Send previous track command."""
-        self.device.skipPrevious(self._active_media_plexapi_type)
+        if self.device:
```
`self.device` is not enough to know if it supports remote commands - needs to include `playback` in the `device.protocolCapabilities` (which still doesn't tell you if it supports volume controls.

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
+    # pylint: disable=W0613
+    def play_media(self, media_type, media_id, **kwargs):
+        """Play a piece of media."""
+        if self.device:
```
`self.device` is not enough to know if it supports remote commands - needs to include `playback` in the `device.protocolCapabilities` (which still doesn't tell you if it supports volume controls.

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
+
+    def _client_play_media(self, media, **params):
+        """Instruct Plex client to play a piece of media."""
+        if self.device:
```
`self.device` is not enough to know if it supports remote commands - needs to include `playback` in the `device.protocolCapabilities` (which still doesn't tell you if it supports volume controls.

balloob replied:
Also guard clause

Teagan42 commented:
I like all the extra details you're adding - but unfortunately is not quite right. I put in quite a few comments that need to be addressed.

Teagan42 commented:
The duplication of entities is the most annoying thing on this, I have 4 copies of one device and 2 of another, that even I was not getting before I added the `machineIdentifier is not None` check in mine.

This is all the same physical device
<img width="339" alt="screen shot 2017-03-12 at 9 49 42 am" src="https://cloud.githubusercontent.com/assets/2989925/23833859/453d638a-0709-11e7-9dad-4b8c83d9a2b0.png">

Same with this:
<img width="344" alt="screen shot 2017-03-12 at 9 50 20 am" src="https://cloud.githubusercontent.com/assets/2989925/23833862/559b7a6e-0709-11e7-9801-a3dcbb670dae.png">

Teagan42 commented:
The request will fail, I have not seen a protocol capability that dictates
it can accept volume commands.

On Mar 12, 2017 11:03 AM, "JesseWebDotCom" <notifications@github.com> wrote:

> *@JesseWebDotCom* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/media_player/plex.py
> <https://github.com/home-assistant/home-assistant/pull/6054#discussion_r105562719>
> :
>
> >
>      def set_volume_level(self, volume):
>          """Set volume level, range 0..1."""
> -        self.device.setVolume(int(volume * 100),
> -                              self._active_media_plexapi_type)
> +        if self.device:
>
> Then there is nothing we can do?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/6054#discussion_r105562719>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC2fZfDsaq9t-QHPTF33jaEImd6o2hAbks5rlCVYgaJpZM4MD1_B>
> .
>

Teagan42 commented:
Also, coveralls is failing because you're branch is too far behind `dev` - do a rebase against `home-assistant/dev` and it should (keyword here) fix it

Teagan42 commented:
@JesseWebDotCom Fix some of these and I'll test it on my setup and review changes again.

JesseWebDotCom commented:
I tried to rebase but am now getting this message when I try to push:
```
Jesse’s MacBook Pro:home-assistant jessetorres$ git push origin HEAD
error: unable to push to unqualified destination: HEAD
The destination refspec neither matches an existing ref on the remote nor
begins with refs/, and we are unable to guess a prefix based on the source ref.
error: failed to push some refs to 'https://github.com/JesseWebDotCom/home-assistant.git'
```

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -7,32 +7,77 @@
 
-SUPPORT_PLEX = SUPPORT_PAUSE | SUPPORT_PREVIOUS_TRACK | SUPPORT_NEXT_TRACK | \
-    SUPPORT_STOP | SUPPORT_VOLUME_SET | SUPPORT_PLAY
+requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
```
Please remove this. You are making all of Home Assistant less secure with this line. We should always notify the user when something shady is going on with certificates

robbiet480 replied:
So this may be a problem or something we have to think through more. I hit this same problem back when I tried getting PMS to connect via SSL. Full notes are [here](https://github.com/home-assistant/home-assistant/issues/891#issuecomment-275016420), I suggest reading them

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -62,6 +107,37 @@ def config_from_file(filename, config=None):
     """Setup the Plex platform."""
+    # optional parameters
+    optional_config = {}
+    optional_config["entity_namespace"] = config.get("entity_namespace")
```
This pretty much seems a mapping of config to a new dict called `optional_config`. Why not just use `config` ?

JesseWebDotCom replied:
Correct - I did that because there is a point in the component where the config dictionary is overwritten with data pulled from plex.conf file.  I'm adding code to just merge the original config dictionary with the one pulled from plex.conf.  Don't know if thats the right way to do it but it works.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -62,6 +107,37 @@ def config_from_file(filename, config=None):
+
+    max_frozen_playing = DEFAULT_MAX_FROZEN_PLAYING
+    try:
+        max_frozen_playing = int(config.get(CONF_MAX_FROZEN_PLAYING))
```
Use voluptuous to check this.

```python
vol.Optional(CONF_MAX_FROZEN_PLAYING, default=DEFAULT_MAX_FROZEN_PLAYING): vol.All(vol.Coerce(int), vol.Range(min=0, max=X),
```

JesseWebDotCom replied:
Thanks - will do

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -62,6 +107,37 @@ def config_from_file(filename, config=None):
+        max_frozen_playing = int(config.get(CONF_MAX_FROZEN_PLAYING))
+    except ValueError:
+        pass
+    finally:
```
This is also wrong. If `config.get(CONF_MAX_FROZEN_PLAYING)` was a string, you will now run into bad code because finally is always executed although coercion failed.

JesseWebDotCom replied:
Removed

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +253,377 @@ def update_sessions():
+        self._frozen_fixed_state = None
+        self._is_player_active = False
+        self._is_player_available = False
+        self._is_player_available = False
```
duplicate?

JesseWebDotCom replied:
Good catch, removed

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +253,377 @@ def update_sessions():
+                self._device.title) or DEVICE_DEFAULT_NAME
+
+        # set valid session, preferring device session
+        if self._device and self.plex_sessions.get(
```
You're fetching the session twice. Once in the if statement and once afterwards.

JesseWebDotCom replied:
I'm not sure how to rewrite this.  I'm basically saying " if we have a device and a device session, then replace self._session with this device session (otherwise leave self._session to what it was previously set to)"

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +253,377 @@ def update_sessions():
+                self._device.machineIdentifier, None))
+
+        if self._session:
+            self._media_position = self._convert_na_to_none(
```
We should not update the position on every update. Instead, you should store the start value when it starts playing and store when you stored that base value and return that as `media_position_updated_at`. Otherwise every time Home Assistant updates this device it will cause a state change and that's no bueno.

JesseWebDotCom replied:
I updated the code to follow the pattern I see the Apple TV component took.  Let me know if that's good enough.

balloob replied:
The AppleTV is pushed based. You are now still updating the media position on every update.

You can see this in the logs. When you are playing some media, you should not see a state change every 10 seconds.

balloob replied:
If it's too complicated to add now, let's remove this functionality for now and keep it for a future PR.

JesseWebDotCom replied:
A few things:
- each plexdevice object created needs to access some of the optional config variables I added.  So I minimally need to pass the config to the object.  I see a line where the config object is being cleared and reset to config grabbed from plex.conf.  So I merged the 2 objects instead of having it be overwritten.  Let me know if that's necessary.
- can you confirm my understanding of what you are suggesting for positioning is correct? I think you are suggesting to fake it by polling the position Initally and then increment the position based on elapsed time (no more polling the position) plus update the position when you see a play, pause, stop, next/previous track event.  I would also only set the _updated_at variable when I actually poll for position not when I fake update the position.  Correct?  If so, my frozen detection feature is almost worthless as I won't be able to quickly detect a plex client who is in a playing or paused state but the position hasn't moved.  For example, start streaming from an iPhone or browser plex client, watch it show as playing in hass, and then kill that client (ex user locks phone, user closes the browser, plex client crashes, etc).  With the way I have the code now, I can know pretty quickly when that happens.  If I change it to my understanding of how you want it, I'll only know when my fake positioning goes past duration.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +253,377 @@ def update_sessions():
+
+        if self._is_player_available:
+            if ((frozen_seconds >
+                 int(self.optional_config[CONF_MAX_FROZEN_PAUSED])) and
```
No need to cast this, Voluptuous should take care of that.

JesseWebDotCom replied:
updated

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +253,377 @@ def update_sessions():
+            self._media_title = self._convert_na_to_none(self._session.title)
+
+        # Movies
+        if (self.media_content_type is MEDIA_TYPE_VIDEO and
```
Please use `==` for comparison

JesseWebDotCom replied:
done

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -157,99 +253,377 @@ def update_sessions():
+
+        # Music
+        if (self._is_player_active and
+                self._media_content_type is MEDIA_TYPE_MUSIC):
```
`==`

JesseWebDotCom replied:
done

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
+    def media_position_updated_at(self):
+        """When was the position of the current playing media valid."""
+        if self._is_player_active and self.state == STATE_PLAYING:
+            self._media_position_updated_at = util.dt.utcnow()
```
Fetching a property should not have side effects like updating a variable. Like I wrote above, you only need to change media_position when the device started/stopped playing.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
+        - On mute, store volume and set volume to 0
+        - On unmute, set volume to previously stored volume
+        """
+        if self.device:
```
Please use guard clauses.

```python
if not self.device:
    return

self._volume_muted = …
```

JesseWebDotCom replied:
done, thanks - never heard of this until you mentioned

balloob reviewed (changes requested):
Already a lot better. Still some work left to do.

JesseWebDotCom committed:
Balloob recommendations and Plex Protocol Capabilities checks

JesseWebDotCom commented:
@Teagan42 - see if I addressed the issues you raised

Teagan42 commented:
Will . check tomorrow

On Mon, Mar 13, 2017 at 6:56 PM, JesseWebDotCom <notifications@github.com>
wrote:

> @Teagan42 <https://github.com/Teagan42> - see if I addressed the issues
> you raised
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/6054#issuecomment-286298550>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC2fZbmqsxg0czpThyU3BBf9k6PdmV0zks5rlfOygaJpZM4MD1_B>
> .
>

robbiet480 commented:
@JesseWebDotCom Did you a favor and rebased on latest dev! Saw you got that permissions change done, sorry I didn't get to it myself :(

robbiet480 committed:
Remove deprecated disable-msg in favor of disable

JesseWebDotCom commented:
@robbiet480 - thanks!  If I need to make more changes, what git commands do I need to run to get your latest changes?

robbiet480 commented:
@JesseWebDotCom just a `git pull` should be all you need

JesseWebDotCom commented:
Thanks, really.  So what are the next steps to get this approved?

Teagan42 commented:
I would like to test this change before any merge.

On Mar 14, 2017 12:58 AM, "JesseWebDotCom" <notifications@github.com> wrote:

> Thanks, really. So what are the next steps to get this approved?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/6054#issuecomment-286338141>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC2fZW6z4EWJef22YPPzHlxKDkwAmwegks5rljqpgaJpZM4MD1_B>
> .
>

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -62,10 +101,13 @@ def config_from_file(filename, config=None):
-    if len(config):
+    # get config from plex.conf
+    file_config = config_from_file(hass.config.path(PLEX_CONFIG_FILE))
+    config.update(file_config)  # merge configs
```
We should not mutate the passed in config. I also think that it's kinda weird, because the config in the file contains ip adresses + tokens ? 

To copy the config, do:

```python
config = dict(config)
```

balloob replied:
Can you explain why we should even consider merging these two configs?

JesseWebDotCom replied:
each plexdevice object created needs to access some of the optional config variables I added.  So I minimally need to pass the config to the object.  I see a line where the config object is being cleared and reset to config grabbed from plex.conf.  So I merged the 2 objects instead of having it be overwritten.  Let me know if that's necessary.

balloob replied:
If an object needs access to both, pass poth objects in. We should not merge the two data sources.

balloob commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +614,316 @@ def _active_media_plexapi_type(self):
+    def media_position_updated_at(self):
+        """When was the position of the current playing media valid."""
+        if self._is_player_active:
+            return util.dt.utcnow()
```
See my above comment. This is not right.

JesseWebDotCom replied:
can you confirm my understanding of what you are suggesting for positioning is correct? I think you are suggesting to fake it by polling the position Initally and then increment the position based on elapsed time (no more polling the position) plus update the position when you see a play, pause, stop, next/previous track event.  I would also only set the _updated_at variable when I actually poll for position not when I fake update the position.  Finally, I can stop fake determining the position once I reach and or exceed duration.  Correct?  If so, my frozen detection feature is almost worthless as I won't be able to quickly detect a plex client who is in a playing or paused state but the position hasn't moved.  For example, start streaming from an iPhone or browser plex client, watch it show as playing in hass, and then kill that client (ex user locks phone, user closes the browser, plex client crashes, etc).  With the way I have the code now, I can know pretty quickly when that happens.  If I change it to my understanding of how you want it, I'll only know when my fake positioning goes past duration.

JesseWebDotCom replied:
"plus update the position when you see a play, pause, stop, next/previous track event" - means I truly grab the position from the client

balloob replied:
We don't keep track of a fake position. We only keep track of a position combined with a timestamp when that was recorded. We update this position as little as possible. Any UI will have to compute the actual current state when it is displaying it.

If you are playing media, it will be set once to 0 at the beginning of the media item and that's it. If Plex stops playing when the media is over, the state will be updated to "PAUSED" or whatever because you're polling for updates.

To be clear: this is the only allowed implementation of `media_position` and `media_position_updated_at`. Any other implementation will not be merged in as it spams the database of our users too much.

JesseWebDotCom replied:
Ok, I'll probably take your suggestion and yank this (position tracking plus my frozen detection feature) out for now.  I'll do that and update the documentation later this evening.  Thanks for the explanation and being patient.

balloob reviewed (changes requested):
Found a few more things that require attention.

JesseWebDotCom committed:
Removed position related items (seek, frozen detection, etc)

JesseWebDotCom commented on homeassistant/components/media_player/plex.py:
```diff
@@ -7,33 +7,63 @@
 import logging
 import os
-from datetime import timedelta
+from datetime import datetime, timedelta
```
Amended

JesseWebDotCom committed:
Removed unused datetime

JesseWebDotCom commented:
So I removed anything position related and updated the documentation.

Teagan42 commented:
I'm still getting a session based client and a regular client from the `server.clients`.

JesseWebDotCom commented:
@Teagan42 - what's your recommendation?

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -272,93 +559,301 @@ def _active_media_plexapi_type(self):
+                        SUPPORT_TURN_OFF | SUPPORT_VOLUME_MUTE)
+
+        # only show controls when we know what device is connecting
+        if not self._make:
```
Should this be an `elif`? It's overwriting the two `if`s before this

JesseWebDotCom replied:
good catch, those are missing return statements - will add

Teagan42 commented on homeassistant/components/media_player/plex.py:
```diff
@@ -136,12 +170,31 @@ def update_devices():
+        if config.get(CONF_INCLUDE_NON_CLIENTS):
+            for machine_identifier, session in plex_sessions.items():
+                if (machine_identifier not in plex_clients
+                        and machine_identifier is not None):
```
This should be to prevent dupes:

` and session.player.machineIdentifier is not None):`

Teagan42 replied:
Tested this change locally - and it prevents dupes.

JesseWebDotCom replied:
I'll add, test, and check in (for this one spot) - thanks

Teagan42 replied:
I think

Teagan42 replied:
Yup this fixes dupe issue

Teagan42 replied:
Ok so this is odd - at first load, with this change, I don't get dupes, but after it sits for a bit, I get dupes... let me investigate more...

Teagan42 commented:
@JesseWebDotCom I'm not sure what is causing duplicate clients, I'm still digging.

JesseWebDotCom commented:
@Teagan42 - Standing by

Teagan42 commented:
So, I added the machine identifier to the state attributes - when it first loads, only the first two show up (yay), but after another pass, the second two show up (empty machine identifiers)
```
media_player.my_afts	idle	machineIdentifer: ead79392dc1a30ac-com-plexapp-android
app_name: 
friendly_name: My AFTS
media_player.my_afts_2	idle	machineIdentifer: 48d04c6c0bf62ed0-com-plexapp-android
app_name: 
friendly_name: My AFTS
media_player.my_afts_3	idle	supported_features: 20533
friendly_name: My AFTS
media_player.my_afts_4	idle	supported_features: 20533
friendly_name: My AFTS
```

So, some how, we're getting empty machine identifiers into the plex clients. Not sure where, again still digging - Not sure how no one else is seeing this issue...

Teagan42 commented:
Even more interesting, realized I don't have the include non clients flag set in my config...

Teagan42 commented:
@JesseWebDotCom if you got time in about an hour we can hop on a video call and I can show you what I'm seeing.

JesseWebDotCom commented:
Unfortunately I'm at a family event.  Perhaps you can send me a clip.  Also, what's your setup (I think I have every plex client in the world - maybe I can reproduce the issue)?

Teagan42 commented:
I am having issues with my Amazon Fire TVs (haven't even checked the two sticks I have up stairs).

On first pass, the AFTVs are added (as expected), but second pass (the refresh) it seems to duplicate them as non-devices.

When I turn on the `INCLUDE_NON_CLIENT` config, those work just fine, no problems.

Teagan42 commented:
TV (non-client) works, Chrome (client) works, AFTVs - duplicated... there's something I am missing somewhere...

Teagan42 commented:
Fire Sticks have same issue as the TVs

Teagan42 commented:
Top to bottom:
Amazon Fire TV A (Duplicated - no machine identifier)
Amazon Fire TV B (Duplicated - no machine identifier)
Amazon Fire TV A (Machine Identifier)
Amazon Fire TV B (Machine Identifier)
Android Phone (Machine Identifier)
Chrome (Machine Identifier)
Chrome (Duplicated - no machine identifier)
Amazon Fire Stick A (Machine identifier)
Amazon Fire Stick A (Duplicated - no machine identifier)
Samsung TV A (Non-client, machine identifier)
Samsung TV B (Non-client, machine identifier)

<img width="142" alt="screen shot 2017-03-15 at 6 21 12 pm" src="https://cloud.githubusercontent.com/assets/2989925/23976262/33b43134-09ac-11e7-8a3e-848dd90c975f.png">

Teagan42 commented:
I have other things to work on tonight - if you can dig in, let me know otherwise I'll try to look tomorrow

JesseWebDotCom commented:
When you see the duplicate devices, go to http://IP:PORT/status/sessions (ip = plex server) and post the results (make sure it doesn't include any sensitive information).

JesseWebDotCom commented:
Also, what do you have for a plex server?

balloob commented:
So the code looks good. This can be merged once the bugs found by @Teagan42 fixed or avoided.

Teagan42 commented:
Bugs were all on my setup, my bad. All is good.

On Mar 15, 2017 11:04 PM, "Paulus Schoutsen" <notifications@github.com>
wrote:

> So the code looks good. This can be merged once the bugs found by
> @Teagan42 <https://github.com/Teagan42> fixed or avoided.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/6054#issuecomment-286958975>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC2fZV20QCbxBLwkEnGni3Km81yYj6Ykks5rmMLqgaJpZM4MD1_B>
> .
>

balloob commented:
Thanks @JesseWebDotCom for the work and @Teagan42 for testing! 🎉  💃  🐬  🍺

JesseWebDotCom commented:
Awesome, thanks for teaching

robbiet480 commented:
@JesseWebDotCom Thanks for hanging on! We thought that this was such a great pull request and show of effort it was deserving of its own [tweet!](https://twitter.com/home_assistant/status/842408632176246784). Cherish it, we don't hand those out lightly 😳

JesseWebDotCom commented:
Thanks.  Hey, I believe I have one more change to make which is a fix to the new show all controls option (it won't work as is).  Can I still submit that as a new commit here or is it too late?  Also, how can I tell if my associated documentation pull request is approved and ready to be released at the same time as this component?
